src/plonk_recursion.rs

```diff
@@ -170,67 +220,147 @@ fn verify_all_ipas<C: HaloCurve, InnerC: HaloCurve<BaseField=C::ScalarField>>(
+        let g65 = builder.exp_constant_usize(generator_n, GRID_WIDTH);
+        builder.mul(zeta, g65)
+    }]);
+    let halo_bs = points
+        .into_iter()
+        .map(|p| halo_g_recursive(builder, p, &ipa_challenges))
+        .collect::<Vec<_>>();
+    let halo_b = reduce_with_powers_recursive(builder, &halo_bs, v);
+
+    verify_ipa::<C, InnerC>(
+        builder,
+        proof,
+        c_reduction,
+        reduced_opening,
+        halo_b,
+        u_scaling,
+        ipa_challenges,
+        schnorr_challenge,
+        security_bits,
+    )
 }
 
 /// Verify the final IPA. Return `(u_l, u_r)`, which roughly correspond to `u` and `u^{-1}` in the
 /// Halo paper, respectively.
-fn verify_ipa<C: HaloCurve, InnerC: HaloCurve<BaseField=C::ScalarField>>(
+fn verify_ipa<C: HaloCurve, InnerC: HaloCurve<BaseField = C::ScalarField>>(
     builder: &mut CircuitBuilder<C>,
     proof: &ProofTarget,
-    p: AffinePointTarget,
-    c: Target,
-    x: Target,
+    commitment: AffinePointTarget,
+    value: Target,
+    halo_b: Target,
+    u_scaling: Target,
     ipa_challenges: Vec<Target>,
-    security_bits: usize
+    schnorr_challenge: Target,
+    security_bits: usize,
 ) -> (Vec<Target>, Vec<Target>) {
+    // The H point used for blinding polynomial commitments. Set to H(degree).
+    let pedersen_h = builder.constant_affine_point(hash_usize_to_curve::<InnerC>(
+        1 << ipa_challenges.len(),
+        security_bits,
+    ));
     // Now we begin IPA verification by computing P' and u' as in Protocol 1 of Bulletproofs.
     // In Protocol 1 we compute u' = [x] u, but we leverage to endomorphism, instead computing
     // u' = [n(x)] u.
-    // u is set to H(degree + 1). 
-    let u = builder.constant_affine_point(
-        hash_usize_to_curve::<InnerC>((1 << ipa_challenges.len()) + 1 , security_bits)
-    );
-    let u_prime = builder.curve_mul_endo::<InnerC>(CurveMulOp { scalar: x, point: u }).mul_result;
+    // u is set to H(degree + 1).
+    let u = builder.constant_affine_point(hash_usize_to_curve::<InnerC>(
+        (1 << ipa_challenges.len()) + 1,
+        security_bits,
+    ));
+    let u_prime = builder
+        .curve_mul_endo::<InnerC>(CurveMulOp {
+            scalar: u_scaling,
+            point: u,
+        })
+        .mul_result;
 
     // Compute [c] [n(x)] u = [c] u'.
-    let u_n_x_c = builder.curve_mul::<InnerC>(CurveMulOp { scalar: c, point: u_prime });
-    let p_prime = builder.curve_add::<InnerC>(p, u_n_x_c);
+    let u_n_x_c = builder.curve_mul::<InnerC>(CurveMulOp {
+        scalar: value,
+        point: u_prime,
+    });
+    let p_prime = builder.curve_add::<InnerC>(commitment, u_n_x_c);
 
     // Compute Q as defined in the Halo paper.
     let mut q_muls = Vec::new();
     for i in 0..proof.halo_l_i.len() {
         let l_i = proof.halo_l_i[i];
         let r_i = proof.halo_r_i[i];
-        let l_challenge = ipa_challenges[i];
+        let l_challenge = builder.square(ipa_challenges[i]);
```

Comment:
This doesn't hurt (assuming the prover & other verifier match), but these scalars will be passed through that injective `n` before we take their square roots, so I don't think squaring at this point guarantees that we'll be able to take the square root later.

I haven't gotten around to implementing it yet, but this was my plant to make sure the sqrt below doesn't fail, after discussing a bit with Daira in that Telegram group: when we sample each IPA challenge `x`, we could check if `n(x)` is square, and if not, rewind and sample new randomness for `L_j` / `R_j`. (Challenger doesn't support rewinding currently, but we can make it copy-on-write or something.)

Reply:
I added a TODO for this.
It seems like the loop to check if `n(x)` is square and backtrack if not will be expensive. Maybe there is a way to define another function that maps to squares instead.

Reply:
Also, why can't we use `n(x)` and `n(x)^2` instead of `√n(x)` and `n(x)`?

Reply:
Yeah I think the recursive verifier is enforcing that each `n(ipa_challenge[i])` is square, since IIRC it calls `deterministic_square_root` on those values, and that method checks the square of the purported sqrt. (The verifier isn't computing `n` explicitly, but gets those values implicitly from CurveEndoGate's "signed accumulator" wires.)